feat: Allow fetch functions to pass headers

kadirchan opened the pull request:
This PR opened for issue #1983 

Two exposed method implemented to set default headers to be passed both Mina and Archive GraphQL nodes for fetch operations. Named `setMinaDefaultHeaders`, `setArchiveDefaultHeaders` same as stated in issue. 

Also one new internal method to decide header to be sent to the server, after classify them between archive or not.

In addition optional per request header option parameters added to following methods:
`fetchAccount`
`fetchLastBlock`
`fetchCurrentSlot`
`fetchTransactionStatus`
`sendZkapp`
`fetchEvents`
`fetchActions`
`fetchGenesisConstants`
So that developers can decide between default header usage and one time only usage, and they can overwrite default headers with this usage.

Also unit-tests implemented by using mocking fetch.

kadirchan commented:
I also hesitated to add this to methods `checkZkAppTransaction`, `markAccountToBeFetched` etc., so I skipped them for now.

45930 commented on src/lib/mina/fetch.ts:
```diff
@@ -1049,7 +1128,11 @@ async function makeGraphqlRequest<TDataResponse = any>(
-        headers: { 'Content-Type': 'application/json' },
+        headers: {
+          'Content-Type': 'application/json',
+          ...(isArchiveUrl(url) ? archiveDefaultHeaders : minaDefaultHeaders),
```
I think it's cleaner to pass this function the exact headers we want to use.  That way we can get rid of the `isArchiveUrl` logic.

We know that `fetchTransactionStatus` goes to a Mina node and `fetchEvents` goes to an archive endpoint.  So we can set the correct headers directly in those methods:


```ts
// fetchEvents, for example
let [response, error] = await makeGraphqlRequest<EventQueryResponse>(
    getEventsQuery(
      publicKey,
      tokenId ?? TokenId.toBase58(TokenId.default),
      filterOptions
    ),
    graphqlEndpoint,
    networkConfig.archiveFallbackEndpoints,
    { ...archiveDefaultHeaders, ...headers}
```

45930 commented on src/lib/mina/fetch.ts:
```diff
@@ -102,6 +108,52 @@ function checkForValidUrl(url: string) {
+  archiveDefaultHeaders = headers;
+}
+
 function setGraphqlEndpoints([
```
Together with the `Mina.Network` suggested change, you can make an update here: https://github.com/o1-labs/o1js/blob/e1d70be6d90c88d5be5a91556af17a40f171fadd/src/lib/mina/mina.ts#L139 which would make it super clean.

45930 reviewed (changes requested):
I think you should get rid of the `isArchiveUrl` conditional logic as it's not needed.

I also left a proposal to bake these headers into the `Mina.Network` function, for convenience.  That's not necessary, but I think it could be nice.

I love the tests you added!  Overall this looks great!

kadirchan commented on src/lib/mina/fetch.ts:
```diff
@@ -85,6 +87,10 @@ type ActionsQueryInputs = {
 
+let minaDefaultHeaders: HeadersInit = {};
+
+let archiveDefaultHeaders: HeadersInit = {};
```
Sure

kadirchan commented:
I changed fetch functions again as I understand, removed isArchive and added setting headers feature in `Mina.Netwok` and tested it same way. But there is some other methods that we can add per request header option like `Mina.sendTransaction`, `Mina.fetchEvents`, `Mina.fetchActions`, `Mina.waitForFunding` etc. should I added this option for them too @45930 ?

45930 commented:
@kadirchan 

> But there is some other methods that we can add per request header option like Mina.sendTransaction, Mina.fetchEvents, Mina.fetchActions, Mina.waitForFunding etc.

IMO yes let's add these as well.  At least fetchActions and fetchEvents, because I think the archive node will make more use of this than the mina node.

45930 reviewed (commented):
This looks really good!  I think it's ready to approve if you add a changelog entry, and I will wait to review those final changes from the Mina module you talked about.

kadirchan commented:
Finally added to `Mina.fetchActions`, `Mina.fetchEvents` (also added for `faucet` and `waitForFunding` just in case ). Due to `sendTransaction`  is not directly exposed for devs I didn't add for it but it still uses the default one which is enough imo.

45930 commented:
Thanks @kadirchan !
